jitrebalance/jitrebalance.py

```diff
@@ -0,0 +1,160 @@
+
+    chan = None
+    peer = None
+    for p in peers:
+        for c in p['channels']:
+            if c['short_channel_id'] == scid:
+                chan = c
+                peer = p
+
+    # Check if the channel is active and routable, otherwise there's little
+    # point in even trying
+    if not peer['connected'] or chan['state'] != "CHANNELD_NORMAL":
+        request.set_result({"result": "continue"})
+        return
+
+    # Need to consider who the funder is, since they are paying the fees.
+    funder = chan['msatoshi_to_us_max'] == chan['msatoshi_total']
+    forward_amt = Millisatoshi(onion['per_hop_v0']['forward_amount'])
+    incoming_amt = Millisatoshi(htlc['amount'])
+    fee = incoming_amt - forward_amt
+    pprint(fee)
+    # Compute the amount we need to rebalance, give us a bit of breathing room
+    # while we're at it (25% more rebalancing than strictly necessary) so we
+    # don't end up with a completely unbalanced channel right away again, and
+    # to account for a bit of fuzziness when it comes to dipping into the
+    # reserve.
+    amt = ceil(int(forward_amt - chan['spendable_msat']) * 1.25)
+
+    # If we have a higher balance than is required we don't need to rebalance,
+    # just stop here.
+    if amt <= 0:
+        request.set_result({"result": "continue"})
+        return
+
+    # Compute the last leg of the route first, so we know the parameters to
+    # traverse that last edge.
+    reverse_chan = plugin.rpc.listchannels(scid)['channels']
+    assert(len(reverse_chan) == 2)
+    reverse_chan = [
+        c for c in reverse_chan if c['channel_flags'] != chan['direction']
+    ][0]
+
+    if reverse_chan is None:
+        print("Could not compute parameters for the last hop")
+        request.set_result({"result": "continue"})
+        return
+
+    last_amt = ceil(float(amt) +
+                    float(amt) * reverse_chan['fee_per_millionth'] / 10**6 +
+                    reverse_chan['base_fee_millisatoshi'])
+    last_cltv = 9 + reverse_chan['delay']
+
+    # Exclude the channel we are trying to rebalance when searching for a
+    # path. We will manually append it to the route and bump the other
+    # parameters so it can be used afterwards
+    exclusions = [
+        "{scid}/{direction}".format(scid=scid, direction=chan['direction'])
+    ]
+
+    route = plugin.rpc.getroute(
+        node_id=peer['id'],
+        msatoshi=last_amt,
+        riskfactor=1,
+        exclude=exclusions,
+        cltv=last_cltv,
+    )['route']
+
+    # Append the last hop we computed manually above
+    route += [{
+        'id': plugin.node_id,
+        'channel': scid,
+        'direction': chan['direction'],
+        'msatoshi': amt,
+        'amount_msat': '{}msat'.format(amt),
+        'delay': 9
+    }]
+
+    # We're about to initiate a rebalancing, we'd better remember how we can
+    # settle it once we see it back here.
+
+    payment_key = secrets.token_bytes(32)
```

Comment:
there was the idea floating around to make rebalancing to the same payment hash. but of course this would also confuse the peer as he might have two htlcs in oposing directions and it might make the logic more difficult here

Reply:
With the recent addition of AMP this is not an issue with c-lightning anymore. But plugins may still get some things wrong. The same hash would leak the purpose and allow coalescing the rebalancing with the payment that caused it, but that's a minor issue imho.